File: app/src/main/java/com/haoyu/app/fragment/PageResourcesFragment.java
package com.haoyu.app.fragment;

import android.content.Intent;
import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.view.View;
import android.widget.LinearLayout;

import com.haoyu.app.activity.MFileInfoActivity;
import com.haoyu.app.adapter.PageResourcesAdapter;
import com.haoyu.app.base.BaseFragment;
import com.haoyu.app.entity.CourseResourceListResult;
import com.haoyu.app.entity.MFileInfo;
import com.haoyu.app.entity.Paginator;
import com.haoyu.app.entity.ResourcesEntity;
import com.haoyu.app.lego.student.R;
import com.haoyu.app.utils.Constants;
import com.haoyu.app.utils.OkHttpClientManager;
import com.haoyu.app.view.LoadFailView;
import com.haoyu.app.view.LoadingView;
import com.haoyu.app.xrecyclerview.XRecyclerView;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import okhttp3.Request;

/**
 * 创建日期：2016/11/29 on 17:11
 * 描述:
 * 作者:马飞奔 Administrator
 */
public class PageResourcesFragment extends BaseFragment implements XRecyclerView.LoadingListener {
    @BindView(R.id.loadView)
    LoadingView loadView;
    @BindView(R.id.loadFailView)
    LoadFailView loadFailView;
    @BindView(R.id.li_emptyResources)
    LinearLayout emptyResources; // 空资源
    @BindView(R.id.xRecyclerView)
    XRecyclerView xRecyclerView;
    private boolean isLoadMore = false;
    private boolean isRefresh = false;
    private List<ResourcesEntity> resourcesList = new ArrayList<>();
    private PageResourcesAdapter adapter;
    private String courseId;
    private int limit = 20;
    private int page = 1; // 按页数查询
    private String orders = "CREATE_TIME.DESC";

    @Override
    public int createView() {
        return R.layout.fragment_page_resources;
    }

    @Override
    public void initData() {
        String url = Constants.OUTRT_NET + "/m/resource/ncts?resourceRelations[0].relation.id="
                + courseId + "&resourceRelations[0].relation.type=course" + "&page=" + page
                + "&limit=" + limit + "&orders=" + orders;
        addSubscription(OkHttpClientManager.getAsyn(context, url, new OkHttpClientManager.ResultCallback<CourseResourceListResult>() {
            @Override
            public void onBefore(Request request) {
                if (isRefresh || isLoadMore) {
                    loadView.setVisibility(View.GONE);
                } else {
                    loadView.setVisibility(View.VISIBLE);
                }
            }

            @Override
            public void onError(Request request, Exception e) {
                loadView.setVisibility(View.GONE);
                if (isRefresh) {
                    xRecyclerView.refreshComplete(false);
                } else if (isLoadMore) {
                    xRecyclerView.loadMoreComplete(false);
                    page -= 1;
                } else {
                    xRecyclerView.setVisibility(View.GONE);
                    loadFailView.setVisibility(View.VISIBLE);
                }
            }

            @Override
            public void onResponse(CourseResourceListResult response) {
                loadView.setVisibility(View.GONE);
                if (response != null && response.getResponseData() != null
                        && response.getResponseData().getResources() != null
                        && response.getResponseData().getResources().size() > 0) {
                    updateUI(response.getResponseData().getResources(), response.getResponseData().getPaginator());
                } else {
                    xRecyclerView.setVisibility(View.GONE);
                    emptyResources.setVisibility(View.VISIBLE);
                }
            }
        }));
    }

    private void updateUI(List<ResourcesEntity> resources, Paginator paginator) {
        if (xRecyclerView.getVisibility() != View.VISIBLE)
            xRecyclerView.setVisibility(View.VISIBLE);
        if (isRefresh) {
            resourcesList.clear();
            xRecyclerView.refreshComplete(true);
        } else if (isLoadMore) {
            xRecyclerView.loadMoreComplete(true);
        }
        resourcesList.addAll(resources);
        adapter.notifyDataSetChanged();
        if (paginator != null && paginator.getHasNextPage()) {
            xRecyclerView.setLoadingMoreEnabled(true);
        } else {
            xRecyclerView.setLoadingMoreEnabled(false);
        }
    }

    @Override
    public void initView(View view) {
        Bundle bundle = getArguments();
        if (bundle != null) {
            courseId = bundle.getString("entityId");
        }
        LinearLayoutManager layoutManager = new LinearLayoutManager(context);
        layoutManager.setOrientation(LinearLayoutManager.VERTICAL);
        xRecyclerView.setLayoutManager(layoutManager);
        adapter = new PageResourcesAdapter(context, resourcesList);
        xRecyclerView.setAdapter(adapter);
        xRecyclerView.setLoadingListener(this);
    }

    @Override
    public void setListener() {
        loadFailView.setOnRetryListener(new LoadFailView.OnRetryListener() {
            @Override
            public void onRetry(View v) {
                initData();
            }
        });
        adapter.setCallBack(new PageResourcesAdapter.OpenResourceCallBack() {
            @Override
            public void open(String resourcesName, MFileInfo mFileInfo) {
                mFileInfo.setFileName(resourcesName);
                Intent intent = new Intent(context, MFileInfoActivity.class);
                intent.putExtra("fileInfo", mFileInfo);
                startActivity(intent);
            }
        });
    }

    @Override
    public void onRefresh() {
        isRefresh = true;
        isLoadMore = false;
        page = 1;
        initData();
    }

    @Override
    public void onLoadMore() {
        isRefresh = false;
        isLoadMore = true;
        page += 1;
        initData();
    }
}

File: app/src/main/java/com/haoyu/app/fragment/PageResourcesFragment.kt
package com.haoyu.app.fragment

import android.content.Intent
import android.support.v7.widget.LinearLayoutManager
import android.view.View
import android.widget.TextView
import com.haoyu.app.activity.MFileInfoActivity
import com.haoyu.app.adapter.PageResourcesAdapter
import com.haoyu.app.base.BaseFragment
import com.haoyu.app.entity.CourseResources
import com.haoyu.app.entity.Paginator
import com.haoyu.app.entity.ResourcesEntity
import com.haoyu.app.lego.student.R
import com.haoyu.app.utils.Constants
import com.haoyu.app.utils.OkHttpClientManager
import com.haoyu.app.view.LoadFailView
import com.haoyu.app.view.LoadingView
import com.haoyu.app.xrecyclerview.XRecyclerView
import okhttp3.Request
import java.util.*

/**
 * 创建日期：2018/1/15.
 * 描述:课程学习资源列表
 * 作者:xiaoma
 */
class PageResourcesFragment : BaseFragment(), XRecyclerView.LoadingListener {
    private lateinit var loadingView: LoadingView
    private lateinit var loadFailView: LoadFailView
    private lateinit var tvEmpty: TextView
    private lateinit var xRecyclerView: XRecyclerView
    private val mDatas = ArrayList<ResourcesEntity>()
    private lateinit var adapter: PageResourcesAdapter
    private var courseId: String? = null
    private var isLoadMore = false
    private var isRefresh = false
    private var page = 1
    override fun createView(): Int {
        return R.layout.fragment_page_resources
    }

    override fun initView(view: View) {
        courseId = arguments?.getString("entityId")
        loadingView = view.findViewById(R.id.loadingView)
        loadFailView = view.findViewById(R.id.loadFailView)
        tvEmpty = view.findViewById(R.id.tv_empty)
        xRecyclerView = view.findViewById(R.id.xRecyclerView)
        val layoutManager = LinearLayoutManager(context)
        layoutManager.orientation = LinearLayoutManager.VERTICAL
        xRecyclerView.layoutManager = layoutManager
        adapter = PageResourcesAdapter(context, mDatas)
        xRecyclerView.adapter = adapter
        xRecyclerView.setLoadingListener(this)
    }

    override fun initData() {
        val url = (Constants.OUTRT_NET + "/m/resource/ncts?resourceRelations[0].relation.id=" + courseId + "&resourceRelations[0].relation.type=course" + "&page=" + page
                + "&limit=20" + "&orders=CREATE_TIME.DESC")
        addSubscription(OkHttpClientManager.getAsyn(context, url, object : OkHttpClientManager.ResultCallback<CourseResources>() {
            override fun onBefore(request: Request) {
                if (isRefresh || isLoadMore) {
                    loadingView.visibility = View.GONE
                } else {
                    loadingView.visibility = View.VISIBLE
                }
            }

            override fun onError(request: Request, e: Exception) {
                loadingView.visibility = View.GONE
                when {
                    isRefresh -> xRecyclerView.refreshComplete(false)
                    isLoadMore -> {
                        page -= 1
                        xRecyclerView.loadMoreComplete(false)
                    }
                    else -> {
                        xRecyclerView.visibility = View.GONE
                        loadFailView.visibility = View.VISIBLE
                    }
                }
            }

            override fun onResponse(response: CourseResources?) {
                loadingView.visibility = View.GONE
                if (response?.getResponseData() != null && response.getResponseData().resources.size > 0) {
                    updateUI(response.getResponseData().resources, response.getResponseData().paginator)
                } else {
                    when {
                        isRefresh -> xRecyclerView.refreshComplete(true)
                        isLoadMore -> xRecyclerView.loadMoreComplete(true)
                        else -> {
                            xRecyclerView.visibility = View.GONE
                            tvEmpty.visibility = View.VISIBLE
                        }
                    }
                }
            }
        }))
    }

    private fun updateUI(resources: List<ResourcesEntity>, paginator: Paginator?) {
        if (xRecyclerView.visibility != View.VISIBLE)
            xRecyclerView.visibility = View.VISIBLE
        if (isRefresh) {
            mDatas.clear()
            xRecyclerView.refreshComplete(true)
        } else if (isLoadMore) {
            xRecyclerView.loadMoreComplete(true)
        }
        mDatas.addAll(resources)
        adapter.notifyDataSetChanged()
        xRecyclerView.isLoadingMoreEnabled = paginator != null && paginator.hasNextPage
    }

    override fun setListener() {
        loadFailView.setOnRetryListener { initData() }
        adapter.setCallBack { resourcesName, mFileInfo ->
            mFileInfo.fileName = resourcesName
            val intent = Intent(context, MFileInfoActivity::class.java)
            intent.putExtra("fileInfo", mFileInfo)
            startActivity(intent)
        }
    }

    override fun onRefresh() {
        isRefresh = true
        isLoadMore = false
        page = 1
        initData()
    }

    override fun onLoadMore() {
        isRefresh = false
        isLoadMore = true
        page += 1
        initData()
    }

}